torch/distributed/fsdp/flat_param.py

```diff
@@ -832,6 +907,41 @@ def prepare_gradient(self):
                 )
             flat_param.grad = None
 
+    def prepare_gradient_for_optim(self):
+        """
+        Prepares the gradient for optimizer computation by moving the sharded
+        gradient to the ``.grad`` attribute.
+        """
+        flat_param = self.flat_param
+        # TODO (awgu): We should replace these conditional checks to encode
+        # the logical intention more directly.
+        if hasattr(flat_param, "_cpu_grad"):
+            # NOTE: This branch includes `NO_SHARD`.
+            self._check_sharded(flat_param)
+            self._check_on_cpu(flat_param)
+            flat_param.grad = flat_param._cpu_grad  # type: ignore[attr-defined]
+        elif hasattr(flat_param, "_saved_grad_shard"):
+            self._check_sharded(flat_param)
+            self._check_on_compute_device(flat_param)
+            self._check_on_compute_device(flat_param._saved_grad_shard)  # type: ignore[attr-defined]
+            # If no sharded gradient was computed this iteration, then there is
+            # no need to forward `_saved_grad_shard` to `grad`
+            if flat_param._post_backward_called:  # type: ignore[attr-defined]
+                flat_param.grad = flat_param._saved_grad_shard  # type: ignore[attr-defined]
+                if self._config.keep_low_precision_grads:
+                    assert flat_param.grad is not None  # mypy
+                    flat_param.grad.data = flat_param.grad.to(self._config.param_dtype)
```

Comment:
does this mean none of the mixed precision / post backward called logic works for CPU offload at the moment?

Reply:
Yes, good catch.

For context, I simply moved the logic to `FlatParamHandle` and should have preserved the semantics. (Originally, I did this because `_use_sharded_grad_views()` was called in this method, so it made sense to do it from `FlatParamHandle`, especially given that `prepare_gradients_for_backward()` is here too.)

From master:
https://github.com/pytorch/pytorch/blob/b6bee3c491d3a9a99920fb67203d2bdb8390ac59/torch/distributed/fsdp/fully_sharded_data_parallel.py#L3444-L3465

Reply:
So mixed precision doesn't work with CPU offload? Can we file an issue for this, seems pretty major?